govcd/disk.go

```diff
@@ -0,0 +1,283 @@
+/*
+ * Copyright 2018 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"bytes"
+	"encoding/xml"
+	"fmt"
+	"github.com/vmware/go-vcloud-director/types/v56"
+	"net/http"
+	"net/url"
+)
+
+// Independent disk
+type Disk struct {
+	Disk   *types.Disk
+	client *Client
+}
+
+// Init independent disk struct
+func NewDisk(cli *Client) *Disk {
+	return &Disk{
+		Disk:   new(types.Disk),
+		client: cli,
+	}
+}
+
+// Create an independent disk in VDC
+// Reference: vCloud API Programming Guide for Service Providers vCloud API 30.0 PDF Page 102 - 103,
+// https://vdc-download.vmware.com/vmwb-repository/dcr-public/1b6cf07d-adb3-4dba-8c47-9c1c92b04857/
+// 241956dd-e128-4fcc-8131-bf66e1edd895/vcloud_sp_api_guide_30_0.pdf
+func (vdc *Vdc) CreateDisk(diskCreateParams *types.DiskCreateParams) (*Disk, error) {
+	var err error
+	var execLink *types.Link
+
+	// Find the proper link for request
+	for _, vdcLink := range vdc.Vdc.Link {
+		if vdcLink.Rel == types.RelAdd && vdcLink.Type == types.MimeDiskCreateParams {
+			execLink = vdcLink
+			break
+		}
+	}
+
+	if execLink == nil {
+		return nil, fmt.Errorf("exec link not found")
+	}
+
+	// Parse request URI
+	reqUrl, err := url.ParseRequestURI(execLink.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("error parse URI: %s", err)
+	}
+
+	// Prepare the request payload
+	diskCreateParams.Xmlns = types.NsVCloud
+
+	xmlPayload, err := xml.Marshal(diskCreateParams)
+	if err != nil {
+		return nil, fmt.Errorf("error xml.Marshal: %s", err)
+	}
+
+	// Send Request
+	req := vdc.client.NewRequest(nil, http.MethodPost, *reqUrl, bytes.NewBufferString(xml.Header+string(xmlPayload)))
+	req.Header.Add("Content-Type", execLink.Type)
+	resp, err := checkResp(vdc.client.Http.Do(req))
+	if err != nil {
+		return nil, fmt.Errorf("error create disk: %s", err)
+	}
+
+	// Decode response
+	disk := NewDisk(vdc.client)
+	if err = decodeBody(resp, disk.Disk); err != nil {
+		return nil, fmt.Errorf("error decoding create disk params response: %s", err)
+	}
+
+	// Return the disk
+	return disk, nil
+}
+
+// Update an independent disk
+// 1 Verify that the disk is not attached to a virtual machine.
+// 2 Use newDiskInfo to change update the independent disk.
+// 3 Return task of independent disk update
+// Please verify the independent disk is not connected to any VM before calling this function.
+// If the independent disk is connected to a VM, the task will be failed.
+// Reference: vCloud API Programming Guide for Service Providers vCloud API 30.0 PDF Page 104 - 106,
+// https://vdc-download.vmware.com/vmwb-repository/dcr-public/1b6cf07d-adb3-4dba-8c47-9c1c92b04857/
+// 241956dd-e128-4fcc-8131-bf66e1edd895/vcloud_sp_api_guide_30_0.pdf
+func (d *Disk) Update(newDiskInfo *types.Disk) (Task, error) {
```

Comment:
Function didn't check if newDiskInfo has all mandatory data and they values are correct.

Reply:
I think it should verify by vCloud API, because task of Update will be failed if it is invalid.

Reply:
Yes it's true and yet I would say it is better to fail faster when you know exactly mandatory fields to be expected and provide more meaningful messages than xml parsing and other errors returned by VCD.

Reply:
Checking should be first thing todo - fail fast. Also are name not needed, as documentation says is required field: https://code.vmware.com/apis/287/vcloud#/doc/doc/types/TaskType.html

Reply:
Is name isn't mandatory and shouldn't checked isn't missing?